lib/mixins/dir-mixin.html

```diff
@@ -0,0 +1,171 @@
+     * directly in `ready()` or in HTML.
+     *
+     * Caveats:
+     * - Applications must set `<html dir="ltr">` or `<html dir="rtl">` to sync direction
+     * - Automatic left-to-right or right-to-left styling is sync'd with the `<html>` element only.
+     * - Changing `dir` at runtime is supported.
+     * - Opting out of the global direction styling is permanent
+     *
+     * @mixinFunction
+     * @polymer
+     * @appliesMixin Polymer.PropertyAccessors
+     * @memberof Polymer
+     */
+    Polymer.DirMixin = Polymer.dedupingMixin((base) => {
+
+      if (!OBSERVER) {
+        OBSERVER = new MutationObserver(updateDirection);
+        OBSERVER.observe(document.documentElement, {attributes: true, attributeFilter: ['dir']});
+        takeRecords = function() {
+          OBSERVER.takeRecords();
+          updateDirection();
+        };
+      }
+
+      /**
+       * @constructor
+       * @extends {base}
+       * @implements {Polymer_PropertyAccessors}
+       */
+      const elementBase = Polymer.PropertyAccessors(base);
+
+      /**
+       * @polymer
+       * @mixinClass
+       * @implements {Polymer_DirMixin}
+       */
+      class Dir extends elementBase {
+
+        /**
+         * @override
+         * @suppress {missingProperties} Interfaces in closure do not inherit statics, but classes do
+         */
+        static _processStyleText(is, template, baseURI) {
+          let cssText = super._processStyleText(is, template, baseURI);
+          cssText = this._replaceDirInCssText(cssText);
+          return cssText;
+        }
+
+        /**
+         * Replace `:dir` in the given CSS text
+         *
+         * @param {string} text CSS text to replace DIR
+         * @return {string} Modified CSS
+         */
+        static _replaceDirInCssText(text) {
+          let replacedText = text;
+          replacedText = replacedText.replace(HOST_DIR, HOST_DIR_REPLACMENT);
+          replacedText = replacedText.replace(EL_DIR, EL_DIR_REPLACMENT);
+          if (text !== replacedText) {
+            this.__activateDir = true;
+          }
+          return replacedText;
+        }
+
+        constructor() {
+          super();
+          /** @type {boolean} */
+          this.__autoDirOptOut = false;
+        }
+
+        /**
+         * @suppress {invalidCasts} Closure doesn't understand that `this` is an HTMLElement
+         */
+        ready() {
+          super.ready();
+          this.__autoDirOptOut = /** @type {!HTMLElement} */(this).hasAttribute('dir');
+        }
+
+        /** @suppress {missingProperties} Closure's understanding of prototypes is bad */
+        connectedCallback() {
+          if (elementBase.prototype.connectedCallback) {
+            /** @type {!Polymer_ElementMixin} */(elementBase).prototype.connectedCallback.call(this);
+          }
+          if (this.constructor.__activateDir) {
+            takeRecords();
```

Comment:
Don't think this will fly (same in disconnected). We cannot update *all* elements as takeRecords does when one is connected

Reply:
Ok, just means I'll have to pull document direction on connected.